### PR TITLE
Fix metadata json manifest convention

### LINKED ISSUE #59923: next build fails with any manifest.json file in root of app directory
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/suspicious-lalande-2q96d2

### To Reproduce

Follow these steps https://nextjs.org/docs/app/api-reference/file-conventions/metadata/manifest to add a `manifest.json` or run `yarn build` from the codesandbox.

### Current vs. Expected behavior

Build fails

```
yarn run v1.22.19
$ next build
   ▲ Next.js 14.0.5-canary.26

   Creating an optimized production build  ..Failed to compile.

./node_modules/next/dist/build/webpack/loaders/next-metadata-route-loader.js?page=%2Fmanifest.json%2Froute&isDynamic=0!./app/manifest.json?__next_metadata_route__
Module parse failed: Cannot parse JSON: Unexpected token 'i', "import { N"... is not valid JSON while parsing 'import { NextResponse } from 'next/serve'
File was processed with these loaders:
 * ./node_modules/next/dist/build/webpack/loaders/next-metadata-route-loader.js
You may need an additional loader to handle the result of these loaders.
Error: Cannot parse JSON: Unexpected token 'i', "import { N"... is not valid JSON while parsing 'import { NextResponse } from 'next/serve'
    at JsonParser.parse (/workspace/node_modules/next/dist/compiled/webpack/bundle5.js:28:997642)
    at /workspace/node_modules/next/dist/compiled/webpack/bundle5.js:28:403525
    at processResult (/workspace/node_modules/next/dist/compiled/webpack/bundle5.js:28:399309)
    at /workspace/node_modules/next/dist/compiled/webpack/bundle5.js:28:400370
    at /workspace/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:8727
    at iterateNormalLoaders (/workspace/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:5565)
    at iterateNormalLoaders (/workspace/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:5650)
    at /workspace/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:5879
    at /workspace/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:4271

Import trace for requested module:
./node_modules/next/dist/build/webpack/loaders/next-metadata-route-loader.js?page=%2Fmanifest.json%2Froute&isDynamic=0!./app/manifest.json?__next_metadata_route__


> Build failed because of webpack errors
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
Binaries:
  Node: 20.9.0
  npm: 9.8.1
  Yarn: 1.22.19
  pnpm: 8.10.2
Relevant Packages:
  next: 14.0.5-canary.26
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure, App Router, Metadata (metadata, generateMetadata, next/head)

### Additional context

<sub>[NEXT-2439](https://linear.app/vercel/issue/NEXT-2439/next-build-fails-with-any-manifestjson-file-in-root-of-app-directory)</sub>

### PR DESCRIPTION
**Pull request by huozhi**
### What

Change from processing the file with `next-metatdata-route-loader` directly into passing the file as loader query, and leave an empty resource file for it. This will resolve the error that users were seeing with `manifest.json` convention.

```
Import trace for requested module:
../../../../packages/next/dist/build/webpack/loaders/next-metadata-route-loader.js?page=%2Fmanifest.jso
n%2Froute&isDynamic=0!./app/manifest.json?__next_metadata_route__
getStaticAssetRouteCode page /manifest.json/route this.resourcePath /Users/huozhi/workspace/next.js/tes
t/e2e/app-dir/metadata-json-manifest/app/manifest.json
```

### Why

I looked at the loader process that the final resource processed by webpack is `json!next-metadata-route-loader...`, which means the builtin json loader processing json file after the metadata route loader. I didn't get chance to solve the ordering issue, so I changed the resourcePath to empty "", and pass the file path as query into the loader to avoid json-loader processing it after transpilation.

Fixes #59923

Closes NEXT-2630
Closes NEXT-2439

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: a407ed8702f4c84c7532b0a6d97718b531d8eecc

`TURBOPACK=1 pnpm test test/integration/create-next-app/templates-app.test.ts` (turbopack)
- create-next-app --app > should create TS appDir projects with --ts
- create-next-app --app > should create JS appDir projects with --js
- create-next-app --app > should create JS appDir projects with --js --src-dir
- create-next-app --app > should create Tailwind CSS appDir projects with --tailwind

<details>
<summary>Expand output</summary>

  ● create-next-app --app › should create TS appDir projects with --ts

    FetchError: request to http://localhost:36891/ failed, reason: connect ECONNREFUSED 127.0.0.1:36891

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● create-next-app --app › should create JS appDir projects with --js

    FetchError: request to http://localhost:43133/ failed, reason: connect ECONNREFUSED 127.0.0.1:43133

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● create-next-app --app › should create JS appDir projects with --js --src-dir

    FetchError: request to http://localhost:32823/ failed, reason: connect ECONNREFUSED 127.0.0.1:32823

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● create-next-app --app › should create Tailwind CSS appDir projects with --tailwind

    FetchError: request to http://localhost:40981/ failed, reason: connect ECONNREFUSED 127.0.0.1:40981

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/metadata-manifest | Change |
| - | - | - | - |
| buildDuration | 13.8s | 13.8s | N/A |
| buildDurationCached | 7.4s | 6.7s | N/A |
| nodeModulesSize | 197 MB | 197 MB | N/A |
| nextStartRea..uration (ms) | 423ms | 424ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/metadata-manifest | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.5 kB | 30.5 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 244 B | 241 B | N/A |
| main-HASH.js gzip | 32.1 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 45.4 kB | 45.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/metadata-manifest | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/metadata-manifest | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.2 kB | 4.2 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.56 kB | 6.56 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/metadata-manifest | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/metadata-manifest | Change |
| - | - | - | - |
| index.html gzip | 529 B | 528 B | N/A |
| link.html gzip | 542 B | 540 B | N/A |
| withRouter.html gzip | 525 B | 522 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/metadata-manifest | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.1 kB | 95.1 kB | N/A |
| page.js gzip | 3.05 kB | 3.06 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/metadata-manifest | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 623 B | 624 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/metadata-manifest | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 96.7 kB | 96.7 kB | ✓ |
| app-page-tur..prod.js gzip | 98.5 kB | 98.5 kB | ✓ |
| app-page-tur..prod.js gzip | 92.9 kB | 92.9 kB | ✓ |
| app-page.run...dev.js gzip | 150 kB | 150 kB | ✓ |
| app-page.run..prod.js gzip | 91.4 kB | 91.4 kB | ✓ |
| app-route-ex...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 20.9 kB | 20.9 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.51 kB | 9.51 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.79 kB | 9.79 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.51 kB | 9.51 kB | ✓ |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| pages.runtim...dev.js gzip | 23 kB | 23 kB | ✓ |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| server.runti..prod.js gzip | 50.6 kB | 50.6 kB | ✓ |
| Overall change | 950 kB | 950 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/metadata-manifest | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | ⚠️ +1.35 kB |
| index.pack gzip | 105 kB | 105 kB | ⚠️ +646 B |
| Overall change | 1.66 MB | 1.67 MB | ⚠️ +2 kB |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: ee7fae6c662a5e152b2297e4ab0d4632e2ace0ec

File: packages/next/src/build/webpack/loaders/next-app-loader.ts [hunk 1]
```diff
@@ -124,8 +124,9 @@ async function createAppRouteCode({
 
     resolvedPagePath = `next-metadata-route-loader?${stringify({
       page,
+      filePath: resolvedPagePath,
       isDynamic: isDynamic ? '1' : '0',
-    })}!${resolvedPagePath}${`?${WEBPACK_RESOURCE_QUERIES.metadataRoute}`}`
+    })}!?${WEBPACK_RESOURCE_QUERIES.metadataRoute}`
```
Comment (shuding) [inline packages/next/src/build/webpack/loaders/next-app-loader.ts:129]: Kinda confused by the `!` here... is that still needed as we don't even need a resource path?

Comment (huozhi) [reply packages/next/src/build/webpack/loaders/next-app-loader.ts:129]: We need a `!` to identify this is a loader request otherwise it cannot be resolved

**Review by shuding — APPROVED** [review]

**Pull request merged**